### PR TITLE
Message filtering

### PR DESCRIPTION
**Pull request by Explosion-Scratch**
Adds an addon I made that filters messages! =D
image
image
image

### CODE REVIEW

File: addons/message-filters/userscript.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      remixes: "mod-remix-project",
+      follows: "mod-follow-user",
+      "studio invites": "mod-curator-invite",
+      "forum activity": "mod-forum-activity",
+  }
+  let active = Object.keys(filter).map((i) => filter[i]);
+  let checkboxes = document.createElement("div");
+  checkboxes.classList.add("checkboxes");
+  let heading = document.createElement("h4");
+  heading.appendChild(document.createTextNode("I would like to see notifications from:"));
```
Comment (TheColaber) [inline addons/message-filters/userscript.js:20]: https://github.com/ScratchAddons/ScratchAddons/wiki/Localizing-addons

File: addons/message-filters/userscript.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  checkboxes.appendChild(heading);
+  let keys = Object.keys(filter);
+  for (let i = 0; i < keys.length; i++) {
+      let inp_container = document.createElement("div");
+      inp_container.classList.add("input_container");
+      let inp = document.createElement("input");
+      inp.setAttribute("type", "checkbox");
```
Comment (TheColaber) [inline addons/message-filters/userscript.js:27]: inp.type = 
// I know I'm stubborn.

File: addons/message-filters/userscript.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+              }
+          }
+      }, 50)
+      inp_container.appendChild(inp);
+      let label = document.createElement("label");
+      label.classList.add("input_label");
+      label.setAttribute("for", String.fromCharCode(97 + i));
+      label.appendChild(document.createTextNode(keys[i].replace(keys[i][0], keys[i][0].toUpperCase())));
+      inp_container.appendChild(label);
+      inp_container.setAttribute("style", "padding: 5px; width: 49%; display: inline-block; text-align: left; width: 20vmax; display: inline-block; padding: 10px; border-radius: 5px; transition-duration: .3s; transition-property: background-color; cursor: pointer; overflow: hidden; height: 50%; ")
```
Comment (TheColaber) [inline addons/message-filters/userscript.js:50]: If you are going to have all these styles, why not use make a userstyle?

**Comment by retronbv** [comment]
But does it work?

File: addons/message-filters/userscript.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      inp_container.onmouseover = () => inp_container.style.background = "#05f3";
+      inp_container.onmouseout = () => inp_container.style.background = "transparent";
+      checkboxes.appendChild(inp_container);
+  }
+  setInterval(() => {
```
Comment (TheColaber) [inline addons/message-filters/userscript.js:55]: No setIntervals.

**Review by TheColaber — CHANGES_REQUESTED** [review]

File: addons/message-filters/addon.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    "userscripts": [
+      {
+        "url": "userscript.js",
+        "matches": ["https://scratch.mit.edu/*"]
+      }
+    ],
+    "tags": [],
+    "enabledByDefault": false
+  }
+
```
Comment (TheColaber) [inline addons/message-filters/addon.json:19]: Could also be settings for that addon in the settings page as default options, like onion skinning.

File: addons/message-filters/userscript.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
+export default async function({addon,global,console}) {
+  let list = await addon.tab.waitForElement(".messages-social-list", {
```
Comment (TheColaber) [inline addons/message-filters/userscript.js:2]: Does this work if you load more messages?

**Review by TheColaber — CHANGES_REQUESTED** [review]

File: addons/addons.json [hunk 7]
```diff
@@ -14,6 +14,7 @@
   "// For website:",
   "scratchr2",
   "dark-www",
+  "message-filters",
```
Comment (TheColaber) [inline addons/addons.json:17]: This meant themes for the website. Bring this to the bottom of the list.

File: addons/message-filters/userscript.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      }, 50)
+      inp_container.appendChild(inp);
+      let label = document.createElement("label");
+      label.classList.add("input_label");
+      label.setAttribute("for", String.fromCharCode(97 + i));
+      label.appendChild(document.createTextNode(keys[i].replace(keys[i][0], keys[i][0].toUpperCase())));
+      inp_container.appendChild(label);
+      inp_container.setAttribute("style", "padding: 5px; width: 49%; display: inline-block; text-align: left; width: 20vmax; display: inline-block; padding: 10px; border-radius: 5px; transition-duration: .3s; transition-property: background-color; cursor: pointer; overflow: hidden; height: 50%; ")
+      inp_container.onmouseover = () => inp_container.style.background = "#05f3";
+      inp_container.onmouseout = () => inp_container.style.background = "transparent";
```
Comment (GrahamSH-LLK) [inline addons/message-filters/userscript.js:52]: :hover in css

**Comment by GrahamSH-LLK** [comment]
Turn on Prettier (enable actions at https://github.com/Explosion-Scratch/ScratchAddons/actions) then make a dummy commit

**Comment by mxmou** [comment]
Screenshot

Consistent padding, please.

**Comment by retronbv** [comment]
Why not just make it addon settings?

**Comment by WorldLanguages** [comment]
> Why not just make it addon settings?

This is for filtering messages, it makes sense you might want to filter differently each time

**Comment by retronbv** [comment]
> > Why not just make it addon settings?
> 
> This is for filtering messages, it makes sense you might want to filter differently each time

That is what the addon settings popup is for? Is it not?

File: addons/message-filters/userscript.js (hunk 6 above)
Comment (Explosion-Scratch) [reply addons/message-filters/userscript.js:2]: Yes, that's why I need the setInterval.

File: addons/addons.json (hunk 7 above)
Comment (Explosion-Scratch) [reply addons/addons.json:17]: Got it

File: addons/message-filters/userscript.js (hunk 6 above)
Comment (TheColaber) [reply addons/message-filters/userscript.js:2]: Instead, use an `await addon.tab.waitForElement` for new messages elements.

Comment (RedGuy12) [reply addons/message-filters/userscript.js:2]: Use `while(true)`

**Comment by Explosion-Scratch** [comment]
Still a draft, working on making sure it works with all the new changes

**Comment by Explosion-Scratch** [comment]
> Screenshot
> 
> Consistent padding, please.

Padding is even:
image
It has a hover state.

**Comment by Explosion-Scratch** [comment]
(Speaking of which I'm gonna make the background lighter)

File: addons/message-filters/addon.json (hunk 5 above)
Comment (Explosion-Scratch) [reply addons/message-filters/addon.json]: No, users may want something different easily.

File: addons/message-filters/userscript.js (hunk 6 above)
Comment (Explosion-Scratch) [reply addons/message-filters/userscript.js:2]: I did. Thanks.

**Comment by Hans5958** [comment]
Succeeding #818 cc @TheColaber

**Comment by Explosion-Scratch** [comment]
Mine's better =P

**Comment by Hans5958** [comment]
> Mine's better =P

Yeah I'm just documenting stuff here

**Comment by Explosion-Scratch** [comment]
> > Mine's better =P
> 
> Yeah I'm just documenting stuff here

I was ~~sort of~~ joking 😝

File: addons/message-filters/userscript.js (hunk 1 above)
Comment (Explosion-Scratch) [reply addons/message-filters/userscript.js]: Done!

**Comment by Explosion-Scratch** [comment]
I did l10n, can someone review and check that I did it right?

**Comment by Explosion-Scratch** [comment]
@WorldLanguages?

File: addons/message-filters/addon.json (hunk 5 above)
Comment (TheColaber) [reply addons/message-filters/addon.json]: Not sure you understood what I was trying to tell you. Keep the current settings in the messaging page, but also create some in the settings page as default options. Not sure if you should, I would ask before doing this.

**Comment by Barthdry** [comment]
I think so it is already present in scratch but this one is modified

**Comment by mxmou** [comment]
> Padding is even:
> image
> It has a hover state.

I think it would look better if the title was aligned with the checkboxes themselves, not their containers (which aren't always visible).

**Comment by Hans5958** [comment]
@mxmou 

> I think it would look better if the title was aligned with the checkboxes themselves, not their containers (which aren't always visible).

I think it is fine considering lists/checkboxes like GitHub are like this.
- [x] Notice the margin on the left?

File: addons/message-filters/addon.json (hunk 5 above)
Comment (Explosion-Scratch) [reply addons/message-filters/addon.json]: I think it's fine as is, but what I _do_ want to do is add localStorage saving to it.

**Comment by mxmou** [comment]
> I think it is fine considering lists/checkboxes like GitHub are like this.
> 
> * [x]  Notice the margin on the left?

The margin looks good when there are multiple lists of checkboxes:

Label 1

* [x] Checkbox 1
* [ ] Checkbox 2

Label 2

* [ ] Checkbox 3
* [x] Checkbox 4
* [ ] Checkbox 5

But in this case I think the version with equal padding looks better:

image
image

File: addons/message-filters/style.css [hunk 9]
```diff
@@ -0,0 +1,37 @@
+.checkboxes {
+  box-sizing: border-box;
+  padding: 2vw;
+  width: 100%;
+  background: #05f2;
+  display: block;
+  margin: 0 auto;
+  border-radius: 5px;
+  border: 1px solid #05f3;
+  width: 100%;
```
Comment (mxmou) [inline addons/message-filters/style.css:10]: Duplicates line 4

File: addons/message-filters/style.css [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  margin-top: 10px;
+}
+.checkboxes h4 {
+  padding: 5px;
+}
+.checkboxes input {
+  margin-right: 3px;
+  margin-right: 0.5vmax;
```
Comment (mxmou) [inline addons/message-filters/style.css:18]: There is no reason to specify multiple values for the same property. Only the last will be used.

File: addons/message-filters/style.css [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+}
+.input_container {
+  padding: 5px;
+  width: 49%;
+  display: inline-block;
```
Comment (mxmou) [inline addons/message-filters/style.css:23]: Unnecessary, overridden by lines 25-27.

File: addons/message-filters/style.css (hunk 10 above)
Comment (Explosion-Scratch) [reply addons/message-filters/style.css:18]: Whoops 😝

File: addons/message-filters/style.css (hunk 9 above)
Comment (Explosion-Scratch) [reply addons/message-filters/style.css:10]: Fixed

**Review by TheColaber — APPROVED** [review]

File: addons/message-filters/addon.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "Message filtering",
+  "description": "Enables the ability to finely filter messages on the messages page through checkboxes.",
```
Comment (WorldLanguages) [inline addons/message-filters/addon.json:3]: Addon name and description should make it clear that 1. addon works in messaging page, does not affect other addons that access messages 2. does not exclude the messages you unselect from the message count. I think the best way to do that is to use the words "see" or "hide" (to imply the message count won't be affected) and explicitly say "from the messages page"

**Comment by WorldLanguages** [comment]
Why doesn't clicking here toggle the checkbox?
image

File: addons/message-filters/userscript.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    [msg("studio")]: "mod-studio-activity",
+    [msg("comment")]: "mod-comment-message",
+    [msg("favorite")]: "mod-love-favorite",
+    [msg("love")]: "mod-love-project",
+    [msg("remix")]: "mod-remix-project",
+    [msg("follow")]: "mod-follow-user",
+    [msg("invite")]: "mod-curator-invite",
+    [msg("forum")]: "mod-forum-activity",
+  };
+  let lastTime = 1000;
```
Comment (WorldLanguages) [inline addons/message-filters/userscript.js:17]: Unused variable?

File: addons-l10n/en/message-filters.json [hunk 14]
```diff
@@ -0,0 +1,11 @@
+{
+  "message-filters/notifications": "I would like to see notifications from:",
```
Comment (WorldLanguages) [inline addons-l10n/en/message-filters.json:2]: Use "messages" instead of "notifications"

**Comment by Explosion-Scratch** [comment]
> Why doesn't clicking here toggle the checkbox?
> image

It's the label that has the impact, I'll change the label's style to 100%;

**Comment by mxmou** [comment]
> It's the label that has the impact, I'll change the label's style to 100%;

I don't think `.input_container` and the label need to be separate elements. Couldn't you simply remove the container and make the label have a hover state?

**Comment by WorldLanguages** [comment]
@Explosion-Scratch Please see mxmou's feedback

**Comment by WorldLanguages** [comment]
Might be released in v1.12.0

**Comment by WorldLanguages** [comment]
@Explosion-Scratch Please address mxmou's feedback

**Comment by Explosion-Scratch** [comment]
> @Explosion-Scratch Please address mxmou's feedback

Sorry, been a bit busy. Will do

File: addons/message-filters/userscript.js [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+  // Declare the active array, and either get it from localStorage or set it to every class in filter.
+  // The 'active' array holds the list of the classes of the messages that are shown.
+  let active = JSON.parse(localStorage.getItem("message_preferences")) || Object.keys(filter).map((i) => filter[i]);
+  // Create the checkbox element, which is the container for the message filtering div.
+  let checkboxes = document.createElement("div");
+  // Add a class so I can style it.
+  checkboxes.classList.add("checkboxes");
+  // Create the <h4> element that says (in english) "I would like to see notifications from:"
+  let heading = document.createElement("h4");
+  heading.appendChild(document.createTextNode(msg("messages")));
```
Comment (RedGuy12) [inline addons/message-filters/userscript.js:30]: fyi, `innerText` is fine to use:

```
heading.innerText = msg("messages");
```

**Review by RedGuy12 — APPROVED** [review]

File: addons/message-filters/userscript.js (hunk 15 above)
Comment (Explosion-Scratch) [reply addons/message-filters/userscript.js:30]: OH IT IS?! :facepalm:

Comment (TheColaber) [reply addons/message-filters/userscript.js:30]: Ofc, it's just text.

**Comment by Hans5958** [comment]
Updates?

**Review by RedGuy12 — APPROVED** [review]
pls resolve conflicts so it can be merged

File: addons/addons.json [hunk 16]
```diff
@@ -80,4 +80,6 @@
   "folders",
   "message-filters",
   "block-palette-icons"
```
Comment (RedGuy12) [inline addons/addons.json:82]: missing comma

Comment (mxmou) [reply addons/addons.json:82]: Apparently the merge did something weird. This addon should be at the end of the list.

**Comment by Secret-chest** [comment]
I think we should put another text as the title, maybe "Show messages from:"

**Comment by GrahamSH-LLK** [comment]
@Explosion-Scratch Status?

**Comment by TheColaber** [comment]
He said he will finish in
image

**Comment by GrahamSH-LLK** [comment]
Then I'll attempt to finish this PR when I get home.

**Comment by Explosion-Scratch** [comment]
> He said he will finish in
> image

LEAKED DMS!!!

**Comment by TheColaber** [comment]
> LEAKED DMS!!!

Would you rather my use plain text rather than an image? Or would you rather reply to that comment after 4 days?

File: addons-l10n/en/message-filters.json [hunk 17]
```diff
@@ -0,0 +1,11 @@
+{
+  "message-filters/messages": "I would like to see messages from:",
+  "message-filters/studio": "studio activity",
+  "message-filters/comment": "comments",
+  "message-filters/favorite": "favorites",
+  "message-filters/love": "loves",
+  "message-filters/remix": "remixes",
+  "message-filters/follow": "follows",
+  "message-filters/invite": "studio invite",
```
Comment (Mast3rGenius) [inline addons-l10n/en/message-filters.json:9]: Shouldn't "studio invite" be "studio invites" for consistency's sake?

**Comment by FunctionalMetatable** [comment]
pls merge

File: addons-l10n/en/message-filters.json [hunk 18]
```diff
@@ -0,0 +1,11 @@
+{
+  "message-filters/messages": "I would like to see messages from:",
```
Comment (RedGuy12) [inline addons-l10n/en/message-filters.json:2]: this sounds odd. "I would like to see messages from studio activity"
i think it should be "I would like to see studio activity messages"

Comment (TheColaber) [reply addons-l10n/en/message-filters.json:2]: I think it's fine in the context of the other elements. Plus, we can't just insert "studio activity" into the sentence.

**Comment by WorldLanguages** [comment]
@apple502j Could we move common strings like "loves", "favorites", "studio activity" to _general?

**Comment by Explosion-Scratch** [comment]
@TheColaber @GrahamSH-LLK @RedGuy12 Thanks for all your work on this PR, I appreciate it! (Sorry I sorta abandoned it)

P.S. Thanks for merging WL!)

**Comment by WorldLanguages** [comment]
Sorta? :P

**Pull request merged**